include/partition/cell_storage.hpp

```diff
@@ -331,8 +331,10 @@ template <bool UseShareMemory> class CellStorageImpl
         const auto offset = level_to_cell_offset[level_index];
         const auto cell_index = offset + id;
         BOOST_ASSERT(cell_index < cells.size());
-        return ConstCell{
-            cells[cell_index], weights.data(), source_boundary.data(), destination_boundary.data()};
+        return ConstCell{cells[cell_index],
+                         weights.data(),
+                         source_boundary.empty() ? nullptr : source_boundary.data(),
```

Comment:
All tests that have two  nodes like "a b" will have no source and boundary nodes on all levels. `.data()` or referencing `&..[0]` will an undefined behavior without explicit check for `.empty()`